apisx/v1alpha1/listenerset_types.go

```diff
@@ -0,0 +1,580 @@
+/*
+Copyright 2025 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:categories=gateway-api,shortName=lset
+// +kubebuilder:subresource:status
+// +kubebuilder:storageversion
+// +kubebuilder:printcolumn:name="Accepted",type=string,JSONPath=`.status.conditions[?(@.type=="Accepted")].status`
+// +kubebuilder:printcolumn:name="Programmed",type=string,JSONPath=`.status.conditions[?(@.type=="Programmed")].status`
+// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`
+
+// ListenerSet defines a set of additional listeners
+// to attach to an existing Gateway.
+type ListenerSet struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of ListenerSet.
+	Spec ListenerSetSpec `json:"spec"`
+
+	// Status defines the current state of ListenerSet.
+	//
+	// +kubebuilder:default={conditions: {{type: "Accepted", status: "Unknown", reason:"Pending", message:"Waiting for controller", lastTransitionTime: "1970-01-01T00:00:00Z"},{type: "Programmed", status: "Unknown", reason:"Pending", message:"Waiting for controller", lastTransitionTime: "1970-01-01T00:00:00Z"}}}
+	Status ListenerSetStatus `json:"status,omitempty"`
+}
+
+// ListenerSetSpec defines the desired state of a ListenerSet.
+type ListenerSetSpec struct {
+	// ParentRef references the Gateway that the listeners are attached to.
+	ParentRef ParentGatewayReference `json:"parentRef,omitempty"`
+
+	// Listeners associated with this ListenerSet. Listeners define
+	// logical endpoints that are bound on this referenced parent Gateway's addresses.
+	//
+	// Listeners in a `Gateway` and their attached `ListenerSets` are concatenated
+	// as a list when programming the underlying infrastructure.
+	//
+	// Listeners should be merged using the following precedence:
```

Comment:
When there are conflicts, how are implementations supposed to handle them? IE what if multiple Listeners across different ListenerSets claim the same hostname? Do we reject the entire ListenerSet? Just the Listener? How do we communicate this to users?

Reply:
>  Just the Listener? How do we communicate this to users?

You can reject just the listener and we have a corresponding entry with conditions in the status

Reply:
added a blurb